src/common/btree.c

```diff
@@ -3118,23 +3161,70 @@ btr_tx_tree_destroy(struct btr_context *tcx)
  * The tree open handle is invalid after the destroy.
  *
  * \param toh	[IN]	Tree open handle.
+ * \param args	[IN]	user parameter for btr_ops_t::to_rec_free
  */
 int
-dbtree_destroy(daos_handle_t toh)
+dbtree_destroy(daos_handle_t toh, void *args)
 {
 	struct btr_context *tcx;
+	bool		    destroyed;
 	int		    rc;
 
 	tcx = btr_hdl2tcx(toh);
 	if (tcx == NULL)
 		return -DER_NO_HDL;
 
-	rc = btr_tx_tree_destroy(tcx);
+	D_ASSERT(!tcx->tc_creds_on);
+	rc = btr_tx_tree_destroy(tcx, args, &destroyed);
+	D_ASSERT(rc || destroyed);
 
 	btr_context_decref(tcx);
 	return rc;
 }
 
+/**
+ * This function drains key/values from the tree, each time it deletes a KV
+ * pair, it consumes a @credits, which is input paramter of this function.
+ * It returns if all input credits are consumed, or the tree is empty, in
+ * the later case, it also destroys the btree.
+ *
+ * \param toh		[IN]	 Tree open handle.
+ * \param credis	[IN/OUT] Input and returned drain credits
+ * \param args		[IN]	 user parameter for btr_ops_t::to_rec_free
+ * \param destroy	[OUT]	 Tree is empty and destroyed
+ */
+int
+dbtree_drain(daos_handle_t toh, int *credits, void *args, bool *destroyed)
+{
+	struct btr_context *tcx;
+	int		    rc;
+
+	tcx = btr_hdl2tcx(toh);
+	if (tcx == NULL)
+		return -DER_NO_HDL;
+
+	D_ASSERT(!tcx->tc_creds_on);
+	if (credits) {
+		if (*credits <= 0) {
```

Comment:
The 'tc_creds' is a 30 bits value, so we'd check if 'credits' exceeds max value too.